File: flutter_app/android/wear/src/main/kotlin/systems/neolabs/neorecall/wear/digest/WatchDigest.kt
package systems.neolabs.neorecall.wear.digest

import org.json.JSONArray
import org.json.JSONObject
import systems.neolabs.neorecall.wear.protocol.WearDigestProtocol

/**
 * The phone's account of the day, as the watch is allowed to show it.
 *
 * Parsing is total. A watch that is out of date with its phone must still
 * render: an unknown field is ignored, a missing one falls back to something
 * that draws, and a payload that is not JSON at all yields [empty] rather than
 * an exception on a UI thread the user cannot see.
 */
data class WatchDigest(
  val present: Boolean,
  val signedIn: Boolean,
  val generatedAtMs: Long,
  val receivedAtMs: Long,
  val capture: Capture,
  val today: Today,
  val dayInReview: String?,
  val moment: Moment?,
  val memories: List<Memory>,
  val highlights: List<Highlight>,
) {
  data class Capture(
    val recording: Boolean,
    val title: String,
    val detail: String,
    val pendingSeconds: Int,
    val issue: String?,
  )

  data class Today(
    val talkSeconds: Int,
    val memories: Int,
    val highlights: Int,
    val openTasks: Int,
    val dueToday: Int,
    val overdue: Int,
  )

  /** The newest conversation, with what was said in it. */
  data class Moment(
    val id: String?,
    val title: String?,
    val summary: String?,
    val startedAtMs: Long,
    val endedAtMs: Long,
    val live: Boolean,
    val awaitingWriteUp: Boolean,
    val topics: List<String>,
    val lines: List<Line>,
    /** How many lines the conversation has in total, trimmed or not. */
    val totalLines: Int,
  ) {
    val trimmed: Boolean get() = totalLines > lines.size
  }

  data class Line(val speaker: String?, val text: String, val atMs: Long)

  data class Memory(
    val id: String,
    val emoji: String,
    val title: String,
    val summary: String,
    val typeLabel: String,
    val atMs: Long,
    val highlightCount: Int,
  )

  data class Highlight(
    val id: String,
    val emoji: String,
    val text: String,
    val dueMs: Long?,
    val overdue: Boolean,
    val dueToday: Boolean,
    val memoryTitle: String?,
  )

  /** True once a signed-in phone has sent at least one digest. */
  val usable: Boolean get() = present && signedIn

  /** True when there is something worth opening the digest screen for. */
  val hasContent: Boolean
    get() = usable &&
      (moment != null || memories.isNotEmpty() || highlights.isNotEmpty() ||
        !dayInReview.isNullOrBlank())

  companion object {
    val empty = WatchDigest(
      present = false,
      signedIn = false,
      generatedAtMs = 0L,
      receivedAtMs = 0L,
      capture = Capture(
        recording = false,
        title = "",
        detail = "",
        pendingSeconds = 0,
        issue = null,
      ),
      today = Today(0, 0, 0, 0, 0, 0),
      dayInReview = null,
      moment = null,
      memories = emptyList(),
      highlights = emptyList(),
    )

    fun parse(payload: String?, receivedAtMs: Long): WatchDigest {
      if (payload.isNullOrBlank()) return empty
      val root = try {
        JSONObject(payload)
      } catch (_: Exception) {
        return empty
      }
      return WatchDigest(
        present = true,
        signedIn = root.optBoolean("signedIn", false),
        generatedAtMs = root.optLong("generatedAtMs", 0L),
        receivedAtMs = receivedAtMs,
        capture = capture(root.optJSONObject("capture")),
        today = today(root.optJSONObject("today")),
        dayInReview = root.optString("dayInReview").trimToNull(),
        moment = moment(root.optJSONObject("moment")),
        memories = root.optJSONArray("memories")
          .map(::memory)
          .take(WearDigestProtocol.MAX_MEMORIES),
        highlights = root.optJSONArray("highlights")
          .map(::highlight)
          .take(WearDigestProtocol.MAX_HIGHLIGHTS),
      )
    }

    private fun capture(json: JSONObject?): Capture {
      if (json == null) return empty.capture
      return Capture(
        recording = json.optBoolean("recording", false),
        title = json.optString("title"),
        detail = json.optString("detail"),
        pendingSeconds = json.optInt("pendingSeconds", 0),
        issue = json.optString("issue").trimToNull(),
      )
    }

    private fun today(json: JSONObject?): Today {
      if (json == null) return empty.today
      return Today(
        talkSeconds = json.optInt("talkSeconds", 0),
        memories = json.optInt("memories", 0),
        highlights = json.optInt("highlights", 0),
        openTasks = json.optInt("openTasks", 0),
        dueToday = json.optInt("dueToday", 0),
        overdue = json.optInt("overdue", 0),
      )
    }

    private fun moment(json: JSONObject?): Moment? {
      if (json == null) return null
      val lines = json.optJSONArray("lines")
        .map(::line)
        .filter { it.text.isNotBlank() }
        .take(WearDigestProtocol.MAX_TRANSCRIPT_LINES)
      val title = json.optString("title").trimToNull()
      val summary = json.optString("summary").trimToNull()
      if (lines.isEmpty() && title == null && summary == null) return null
      return Moment(
        id = json.optString("id").trimToNull(),
        title = title,
        summary = summary,
        startedAtMs = json.optLong("startedAtMs", 0L),
        endedAtMs = json.optLong("endedAtMs", 0L),
        live = json.optBoolean("live", false),
        awaitingWriteUp = json.optBoolean("awaitingWriteUp", false),
        topics = json.optJSONArray("topics").strings(),
        lines = lines,
        totalLines = json.optInt("totalLines", lines.size).coerceAtLeast(lines.size),
      )
    }

    private fun line(json: JSONObject) = Line(
      speaker = json.optString("speaker").trimToNull(),
      text = json.optString("text").trim(),
      atMs = json.optLong("atMs", 0L),
    )

    private fun memory(json: JSONObject) = Memory(
      id = json.optString("id"),
      emoji = json.optString("emoji").trimToNull() ?: "💭",
      title = json.optString("title").trimToNull() ?: "Untitled memory",
      summary = json.optString("summary").trim(),
      typeLabel = json.optString("typeLabel").trimToNull() ?: "Moment",
      atMs = json.optLong("atMs", 0L),
      highlightCount = json.optInt("highlightCount", 0),
    )

    private fun highlight(json: JSONObject) = Highlight(
      id = json.optString("id"),
      emoji = json.optString("emoji").trimToNull() ?: "✨",
      text = json.optString("text").trim(),
      dueMs = json.optLongOrNull("dueMs"),
      overdue = json.optBoolean("overdue", false),
      dueToday = json.optBoolean("dueToday", false),
      memoryTitle = json.optString("memoryTitle").trimToNull(),
    )

    private fun <T> JSONArray?.map(transform: (JSONObject) -> T): List<T> {
      if (this == null) return emptyList()
      val items = ArrayList<T>(length())
      for (index in 0 until length()) {
        val row = optJSONObject(index) ?: continue
        items.add(transform(row))
      }
      return items
    }

    private fun JSONArray?.strings(): List<String> {
      if (this == null) return emptyList()
      val items = ArrayList<String>(length())
      for (index in 0 until length()) {
        val value = optString(index).trim()
        if (value.isNotEmpty()) items.add(value)
      }
      return items
    }

    private fun JSONObject.optLongOrNull(key: String): Long? =
      if (has(key) && !isNull(key)) optLong(key) else null

    private fun String?.trimToNull(): String? = this?.trim()?.ifEmpty { null }
  }
}

File: flutter_app/lib/src/watch/watch_digest.dart
import 'dart:convert';

/// Everything a paired Wear OS watch is allowed to show about the day.
///
/// A watch is a display for what the phone already knows: it never queries the
/// server, and nothing it shows can change what the phone does. Keeping the
/// contract in one immutable value means the watch can never render a
/// half-updated mix of two publishes, and keeping it apart from the home-screen
/// widget contract means a field can be added for one surface without paying
/// for it on the other — a Data Layer item is a far tighter budget than a
/// preferences string.
class WatchDigest {
  const WatchDigest({
    required this.signedIn,
    required this.generatedAt,
    required this.capture,
    required this.today,
    this.dayInReview,
    this.moment,
    this.memories = const <WatchDigestMemory>[],
    this.highlights = const <WatchDigestHighlight>[],
  });

  factory WatchDigest.signedOut(DateTime now) => WatchDigest(
    signedIn: false,
    generatedAt: now,
    capture: const WatchDigestCapture(
      recording: false,
      title: '',
      detail: '',
      pendingSeconds: 0,
    ),
    today: const WatchDigestToday(),
  );

  /// Mirrors `WearDigestProtocol` on the Android side. A Data Layer item is
  /// capped near 100 KB and an oversized put fails the whole transfer, so the
  /// payload is trimmed to fit rather than risked.
  static const int maxPayloadBytes = 60000;
  static const int maxTranscriptLines = 40;
  static const int maxLineCharacters = 280;
  static const int maxMemories = 8;
  static const int maxHighlights = 8;
  static const int maxSummaryCharacters = 420;

  final bool signedIn;
  final DateTime generatedAt;
  final WatchDigestCapture capture;
  final WatchDigestToday today;
  final String? dayInReview;
  final WatchDigestMoment? moment;
  final List<WatchDigestMemory> memories;
  final List<WatchDigestHighlight> highlights;

  Map<String, Object?> toJson() => <String, Object?>{
    'signedIn': signedIn,
    'generatedAtMs': generatedAt.millisecondsSinceEpoch,
    'capture': capture.toJson(),
    'today': today.toJson(),
    if (dayInReview != null && dayInReview!.trim().isNotEmpty)
      'dayInReview': dayInReview!.trim(),
    if (moment != null) 'moment': moment!.toJson(),
    'memories': memories.map((memory) => memory.toJson()).toList(),
    'highlights': highlights.map((highlight) => highlight.toJson()).toList(),
  };

  /// The payload as it goes over the Data Layer, guaranteed to fit.
  ///
  /// Trimming drops whole transcript lines from the oldest end first, because a
  /// conversation read on a wrist is read backwards from its most recent line,
  /// and only then gives up list rows. A digest that is refused for being one
  /// byte too large would leave the watch showing yesterday.
  String encode() {
    var candidate = this;
    var payload = jsonEncode(candidate.toJson());
    while (utf8.encode(payload).length > maxPayloadBytes) {
      final reduced = candidate._shrink();
      if (reduced == null) return payload;
      candidate = reduced;
      payload = jsonEncode(candidate.toJson());
    }
    return payload;
  }

  /// One step smaller, or null when there is nothing left to give up.
  WatchDigest? _shrink() {
    final currentMoment = moment;
    if (currentMoment != null && currentMoment.lines.length > 4) {
      final keep = (currentMoment.lines.length * 3) ~/ 4;
      return _copyWith(
        moment: currentMoment.copyWith(
          lines: currentMoment.lines
              .sublist(currentMoment.lines.length - keep)
              .toList(growable: false),
        ),
      );
    }
    if (memories.length > 2) {
      return _copyWith(memories: memories.sublist(0, memories.length - 1));
    }
    if (highlights.length > 2) {
      return _copyWith(
        highlights: highlights.sublist(0, highlights.length - 1),
      );
    }
    if (currentMoment != null && currentMoment.lines.isNotEmpty) {
      return _copyWith(moment: currentMoment.copyWith(lines: const []));
    }
    return null;
  }

  WatchDigest _copyWith({
    WatchDigestMoment? moment,
    List<WatchDigestMemory>? memories,
    List<WatchDigestHighlight>? highlights,
  }) => WatchDigest(
    signedIn: signedIn,
    generatedAt: generatedAt,
    capture: capture,
    today: today,
    dayInReview: dayInReview,
    moment: moment ?? this.moment,
    memories: memories ?? this.memories,
    highlights: highlights ?? this.highlights,
  );
}

/// What capture is doing, in the two lines a watch has room for.
class WatchDigestCapture {
  const WatchDigestCapture({
    required this.recording,
    required this.title,
    required this.detail,
    required this.pendingSeconds,
    this.issue,
  });

  final bool recording;
  final String title;
  final String detail;
  final int pendingSeconds;
  final String? issue;

  Map<String, Object?> toJson() => <String, Object?>{
    'recording': recording,
    'title': title,
    'detail': detail,
    'pendingSeconds': pendingSeconds,
    if (issue != null && issue!.trim().isNotEmpty) 'issue': issue!.trim(),
  };
}

/// The counts the digest's headline row and both complications read from.
class WatchDigestToday {
  const WatchDigestToday({
    this.talkSeconds = 0,
    this.memories = 0,
    this.highlights = 0,
    this.openTasks = 0,
    this.dueToday = 0,
    this.overdue = 0,
  });

  final int talkSeconds;
  final int memories;
  final int highlights;
  final int openTasks;
  final int dueToday;
  final int overdue;

  Map<String, Object?> toJson() => <String, Object?>{
    'talkSeconds': talkSeconds,
    'memories': memories,
    'highlights': highlights,
    'openTasks': openTasks,
    'dueToday': dueToday,
    'overdue': overdue,
  };
}

/// The newest conversation: how it was written up, and what was said in it.
class WatchDigestMoment {
  const WatchDigestMoment({
    this.id,
    this.title,
    this.summary,
    required this.startedAt,
    required this.endedAt,
    required this.live,
    required this.awaitingWriteUp,
    this.topics = const <String>[],
    this.lines = const <WatchDigestLine>[],
    required this.totalLines,
  });

  final String? id;
  final String? title;
  final String? summary;
  final DateTime startedAt;
  final DateTime endedAt;
  final bool live;
  final bool awaitingWriteUp;
  final List<String> topics;
  final List<WatchDigestLine> lines;

  /// How long the conversation is in full, so the watch can say what it is not
  /// showing rather than implying the transcript ends where the payload does.
  final int totalLines;

  WatchDigestMoment copyWith({List<WatchDigestLine>? lines}) =>
      WatchDigestMoment(
        id: id,
        title: title,
        summary: summary,
        startedAt: startedAt,
        endedAt: endedAt,
        live: live,
        awaitingWriteUp: awaitingWriteUp,
        topics: topics,
        lines: lines ?? this.lines,
        totalLines: totalLines,
      );

  Map<String, Object?> toJson() => <String, Object?>{
    if (id != null) 'id': id,
    if (title != null && title!.trim().isNotEmpty) 'title': title!.trim(),
    if (summary != null && summary!.trim().isNotEmpty)
      'summary': summary!.trim(),
    'startedAtMs': startedAt.millisecondsSinceEpoch,
    'endedAtMs': endedAt.millisecondsSinceEpoch,
    'live': live,
    'awaitingWriteUp': awaitingWriteUp,
    'topics': topics,
    'lines': lines.map((line) => line.toJson()).toList(),
    'totalLines': totalLines,
  };
}

/// One thing that was said.
class WatchDigestLine {
  const WatchDigestLine({this.speaker, required this.text, required this.at});

  final String? speaker;
  final String text;
  final DateTime at;

  Map<String, Object?> toJson() => <String, Object?>{
    if (speaker != null && speaker!.trim().isNotEmpty)
      'speaker': speaker!.trim(),
    'text': text,
    'atMs': at.millisecondsSinceEpoch,
  };
}

class WatchDigestMemory {
  const WatchDigestMemory({
    required this.id,
    required this.emoji,
    required this.title,
    required this.summary,
    required this.typeLabel,
    required this.at,
    this.highlightCount = 0,
  });

  final String id;
  final String emoji;
  final String title;
  final String summary;
  final String typeLabel;
  final DateTime at;
  final int highlightCount;

  Map<String, Object?> toJson() => <String, Object?>{
    'id': id,
    'emoji': emoji,
    'title': title,
    'summary': summary,
    'typeLabel': typeLabel,
    'atMs': at.millisecondsSinceEpoch,
    'highlightCount': highlightCount,
  };
}

class WatchDigestHighlight {
  const WatchDigestHighlight({
    required this.id,
    required this.emoji,
    required this.text,
    this.dueAt,
    this.overdue = false,
    this.dueToday = false,
    this.memoryTitle,
  });

  final String id;
  final String emoji;
  final String text;
  final DateTime? dueAt;
  final bool overdue;
  final bool dueToday;
  final String? memoryTitle;

  Map<String, Object?> toJson() => <String, Object?>{
    'id': id,
    'emoji': emoji,
    'text': text,
    if (dueAt != null) 'dueMs': dueAt!.millisecondsSinceEpoch,
    'overdue': overdue,
    'dueToday': dueToday,
    if (memoryTitle != null && memoryTitle!.trim().isNotEmpty)
      'memoryTitle': memoryTitle!.trim(),
  };
}
